examples/cryptocurrency-advanced/frontend/src/plugins/blockchain.js

```diff
@@ -4,201 +4,140 @@ import * as proto from '../../proto/stubs.js'
-                  const bufferWithoutSignature = buffer.subarray(0, buffer.length - 64)
-                  const author = Exonum.uint8ArrayToHexadecimal(buffer.subarray(0, 32))
-                  const signature = Exonum.uint8ArrayToHexadecimal(buffer.subarray(buffer.length - 64, buffer.length));
-
-                  const Transaction = getTransaction(transaction.debug, author)
-
-                  if (Exonum.hash(buffer) !== hash) {
-                    throw new Error('Invalid transaction hash')
-                  }
-
-                  // serialize transaction and compare with message
-                  if (!Transaction.serialize(transaction.debug).every(function (el, i) {
-                    return el === bufferWithoutSignature[i]
-                  })) {
-                    throw new Error('Invalid transaction message')
-                  }
-
-                  if (!Transaction.verifySignature(signature, author, transaction.debug)) {
-                    throw new Error('Invalid transaction signature')
-                  }
-
-                  const transactionData = Object.assign({ hash: hash }, transaction.debug)
-                  if (transactionData.to) {
-                    transactionData.to = Exonum.uint8ArrayToHexadecimal(new Uint8Array(transactionData.to.data))
-                  }
-                  transactions.push(transactionData)
-                }
-
-                return {
-                  block: data.block_proof.block,
-                  wallet: wallet,
-                  transactions: transactions
-                }
-              })
+            .then(({ block_proof, wallet_proof, wallet_history }) => {
+              if (Exonum.verifyBlock(block_proof, validators) !== undefined) throw new Error('Block is invalid')
+              const tableRootHash = Exonum.verifyTable(wallet_proof.to_table, block_proof.block.state_hash, 'crypto.wallets')
+              const walletProof = new Exonum.MapProof(wallet_proof.to_wallet, Exonum.MapProof.rawKey(Exonum.PublicKey), Wallet)
+              if (walletProof.merkleRoot !== tableRootHash) throw new Error('Wallet proof is corrupted')
+
+              const wallet = walletProof.entries.get(Exonum.publicKeyToAddress(publicKey))
+              if (typeof wallet === undefined) throw new Error('Wallet not found')
+
+              const verifiedTransactions = new Exonum.ListProof(wallet_history.proof, Exonum.Hash)
+              const hexHistoryHash = Exonum.uint8ArrayToHexadecimal(new Uint8Array(wallet.history_hash.data))
+              if (verifiedTransactions.merkleRoot !== hexHistoryHash) throw new Error('Transactions proof is corrupted')
+
+              const validIndexes = verifiedTransactions
+                .entries
+                .every(({ index }, i) => i === index)
+              if (!validIndexes) throw new Error('Invalid transaction indexes in the proof')
+
+              const transactions = wallet_history.transactions.map(deserializeWalletTx)
```

Comment:
You still don't check that transactions have correct hashes 🙃 If I understand `deserializeWalletTx` correctly, the check should be something like

```javascript
const correctHashes = transactions.every(({ hash }, i) => verifiedTransactions[i] === hash)
if (!correctHashes) {
  throw new Error('Transaction hash mismatch')
}
```